### ISSUE TITLE
[NEXT-1035] Instrumentation dependencies are not included in standalone output

### ISSUE THREAD

**Issue by owenfarrell** [vercel/next.js#48262]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: x64
      Version: Darwin Kernel Version 22.4.0: Mon Mar  6 21:00:17 PST 2023; root:xnu-8796.101.5~3/RELEASE_X86_64
    Binaries:
      Node: 16.18.1
      npm: 8.19.2
      Yarn: N/A
      pnpm: 8.2.0
    Relevant packages:
      next: 13.3.1-canary.4
      eslint-config-next: N/A
      react: 18.2.0
      react-dom: 18.2.0
```

### Link to the code that reproduces this issue

https://github.com/owenfarrell/nextjs-instrumentation

### To Reproduce

* Create a new Next.JS project using `create-next-app`
* Install a 3rd party dependency for instrumentation purposes (e.g. `applicationinsights`)
* Implement [instrumentation](https://nextjs.org/docs/advanced-features/instrumentation) that references that same 3rd party dependency
* Enable the instrumentation hook and standalone output in `nextjs.config.js`
* Build the project using `next build`

### Describe the Bug

When Next.JS builds a standalone output, the instrumentation source is included a webpack configuration, including nested modules. But the instrumentation source is **_not_** included in output tracing.

As a result, starting the server in standalone mode throws an error when executing the instrumentation hook.

```
Listening on port 3000 url: http://61d4620dca5d:3000
Instrumenting nodejs runtime
Unable to instrument nodejs runtime
Error: Cannot find module 'applicationinsights'
Require stack:
- /app/.next/server/instrumentation.js
- /app/node_modules/next/dist/server/next-server.js
- /app/server.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1075:15)
    at mod._resolveFilename (/app/node_modules/next/dist/build/webpack/require-hook.js:23:32)
    at Module._load (node:internal/modules/cjs/loader:920:27)
    at Module.require (node:internal/modules/cjs/loader:1141:19)
    at require (node:internal/modules/cjs/helpers:110:18)
    at 6691 (/app/.next/server/instrumentation.js:41:54)
    at __webpack_require__ (/app/.next/server/webpack-runtime.js:25:42)
    at Module.register (/app/.next/server/instrumentation.js:19:14)
    at NextNodeServer.prepareImpl (/app/node_modules/next/dist/server/next-server.js:174:85)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/app/.next/server/instrumentation.js',
    '/app/node_modules/next/dist/server/next-server.js',
    '/app/server.js'
  ]
}
```

### Expected Behavior

All dependencies required by the instrumentation hook are included in the standalone output (e.g. `.next/standalon/node_modules`).

### How are you deploying your application? (if relevant)

Docker

<sub>[NEXT-1035](https://linear.app/vercel/issue/NEXT-1035/instrumentation-dependencies-are-not-included-in-standalone-output)</sub>

**Comment by owenfarrell** [comment]
Of note, I've tried a number of attempts at a workaround - specifically:

* Adding the source file(s) to a field under `experimental.outputFileTracingIncludes`

  ...but this option only works for pages, not for code outside of pages (like instrumentation code)

* Adding `applicationinsights` to the list of transpiled modules

  ...but the chunks generated by the list of transpiled modules are not implicitly added to the standalone output
  _**ASIDE: Is that intentional behavior?**_

**Comment by sophiebits** [comment]
I'm seeing a similar issue when deployed to Vercel – server/edge-instrumentation.js is loaded on middleware, but server/instrumentation.js doesn't seem to be loaded at all, which I am suspecting is because it's not correctly included in the .nft.json files.

I tried this patch in hopes that it would fix the issue, but no dice:

```diff
diff --git a/packages/next/src/build/index.ts b/packages/next/src/build/index.ts
index f22fdd1ff..701a8129a 100644
--- a/packages/next/src/build/index.ts
+++ b/packages/next/src/build/index.ts
@@ -1941,6 +1941,14 @@ export default async function build(
                 )
               })
             }
+            if (hasInstrumentationHook) {
+              tracedFiles.add(
+                path.join(
+                  SERVER_DIRECTORY,
+                  `${INSTRUMENTATION_HOOK_FILENAME}.js`
+                )
+              )
+            }
             await promises.writeFile(
               nextServerTraceOutput,
               JSON.stringify({
```

If I'm not mistaken, instrumentation just does not work correctly for the `app` directory. Probably needs some sort of fix similar to #47495.

**Comment by sophiebits** [comment]
Hmm, I think I was seeing something different than this original bug. I fixed my issue in #48557 but I think it was unrelated after all.

**Comment by sophiebits** [comment]
If I'm not mistaken, `copyTracedFiles` needs to call `await handleTraceFiles(path.join(distDir, 'instrumentation.js.nft.json'))` (if hasInstrumentationHook is true).
